PATCH add/remove/replace extensions using path to extension but not to exact attribute

jasonfagerberg-toast opened the pull request:
When you attempt to replace an entire extension schema (using EnterpriseSchema as an example here), scimple throws a Null Pointer exception

```
java.lang.NullPointerException
	at org.apache.directory.scim.core.repository.DefaultPatchHandler.apply(DefaultPatchHandler.java:135)
	at org.apache.directory.scim.core.repository.DefaultPatchHandler.apply(DefaultPatchHandler.java:118)
	at org.apache.directory.scim.core.repository.PatchHandlerTest.applyReplaceEntireEnterpriseExtension(PatchHandlerTest.java:327)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```

This is because it is not treating it as an entire resource replacement but an exact path 

![image](https://github.com/user-attachments/assets/58b134bc-668d-426a-99f8-90e89b7ce31d)

jasonfagerberg-toast committed:
failing test attempting to replace entire enterprise schema

bdemers committed:
Correct new PatchHandlerTest assertion

bdemers committed:
Add ability to accept patches for entire root extension

Fixes: #695

bdemers committed:
Merge branch 'develop' into issue/enterprise-extension-full-repalce

bdemers committed:
[fix] Clean up flow of code

- Removes special checks for operation types, moves logic to specfic PatchOperationHandlers
- Adds test case for removal of full extension

bdemers commented:
@jasonfagerberg-toast I added a fix to this PR.
It's late, and there are a lot of edge cases in this `PatchHandler`, this is one more.

I think you are right, and this is valid patch request, although there are a quark with how the Attribute's are parsed (see comments)

I also added a test case for a "REMOVE" of a full Extension.  I'm not sure if that one is a valid 🤔 
(the containing extension object doesn't have a mutability flag, but it's attributes might, which _could_ make it an illegal operation, I'll have to check with the SCIM working group on that). As is I feel it's better to this code path defined and tested, so it's easier to make changes in the future.

Please take a quick pass through the code and let me know if you think something is missing or it's confusing.  I'll try to take another look when i'm more awake 😴

jasonfagerberg-toast committed:
Added unit test for if path is null but extension is updated in value

jasonfagerberg-toast commented:
> @jasonfagerberg-toast I added a fix to this PR. It's late, and there are a lot of edge cases in this PatchHandler, this is one more.

Wow this is awesome! Wasn't expecting such a quick turn around! From my initial pass this looks good, I found another example online of a PATCH body doing the same thing in a slightly different way: https://is.docs.wso2.com/en/latest/references/extend/user-mgt/provisioning/extend-scim2-user-schemas/

```json
{
   "schemas":[
      "urn:ietf:params:scim:api:messages:2.0:PatchOp"
   ],
   "Operations":[
      {
         "op":"replace",
         "value":{
            "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User":{
               "customAttribute":"new value"
            }
         }
      }
   ]
}
```

So I added a test for that as well, and looks like it is passing 😄

jasonfagerberg-toast commented on scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java:
```diff
@@ -56,6 +57,43 @@ public PatchHandlerTest() {
+    enterpriseExtensionValue.put("department", "New Department");
+    PatchOperation op = patchOperation(REPLACE, enterpriseExtensionUrn, enterpriseExtensionValue);
+    ScimUser updatedUser = patchHandler.apply(user(), List.of(op));
+    EnterpriseExtension actual = (EnterpriseExtension) updatedUser.getExtension("urn:ietf:params:scim:schemas:extension:enterprise:2.0:User");
```
XS Nit:

This one is entirely my fault as I did this when I wrote the test, but we could replace the string here with the variable defined above
```suggestion
    EnterpriseExtension actual = (EnterpriseExtension) updatedUser.getExtension(enterpriseExtensionUrn);
```

jasonfagerberg-toast replied:
Pushed this cleanup to the branch

jasonfagerberg-toast commented on scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java:
```diff
@@ -56,6 +57,43 @@ public PatchHandlerTest() {
+    String enterpriseExtensionUrn = "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User";
+    Map<String, String> enterpriseExtensionValue = new HashMap<>();
+    enterpriseExtensionValue.put("costCenter", "New Cost Center");
+    enterpriseExtensionValue.put("department", "New Department");
```
XS Nit:

Again, entirely my fault, but from reading the rest of the tests, it looks like most use `Map.ofEntries` or `Map.of` to define an immutable map

jasonfagerberg-toast replied:
Pushed this cleanup to the branch

jasonfagerberg-toast committed:
minor test cleanup

jasonfagerberg-toast commented on scim-core/src/main/java/org/apache/directory/scim/core/repository/DefaultPatchHandler.java:
```diff
@@ -132,10 +132,23 @@ private <T extends ScimResource> void apply(T source, Map<String, Object> source
+        patchOperationHandler.applyExtensionValue(source, sourceAsMap, schema, attribute, valuePathExpression, attributeReference.getUrn(), patchOperation.getValue());
+      } else {
+        // If schema is null, it's either the root of an extension, or an invalid patch path
+        // It's not possible from the antlr parser to tell the diff between 'this:is:extension:urn' and 'this:is:extension:urn:attribute'
```
Yeah I was looking at this too, I think the SCIM [RFCs](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2) ANBF has this same issue

```
attrPath  = [URI ":"] ATTRNAME *1subAttr
                 ; SCIM attribute name
                 ; URI is SCIM "schema" URI
``` 

its kinda unfortunate but I think this might be the path of least resistance and path of least hack

jasonfagerberg-toast commented:
> Please take a quick pass through the code and let me know if you think something is missing or it's confusing. I'll try to take another look when i'm more awake 😴

I think for being sleepy this is an excellent PR 😅 I pushed a few small things to the tests but overall LGTM

bdemers commented:
Thanks for taking a look at the updates!

jasonfagerberg-toast commented on scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java:
```diff
@@ -19,6 +19,16 @@
 
 package org.apache.directory.scim.core.repository;
 
+import static java.util.Map.entry;
```
My idea is probably set up with some linters that automatically re-ordered this and remove wildcard imports, I can undo it if you would like!

bdemers replied:
It's fine, it's isolated to a single test class so it's not the end of the world.
But if you want to clean it up or rewrite/squash any of the go for it.

If not, i'll try to merge this later today!

jasonfagerberg-toast committed:
undo import change

bdemers commented:
Sorry about the delay on this one, I though we had already merged it!

Thanks again @jasonfagerberg-toast

jasonfagerberg-toast commented:
> Thanks again @jasonfagerberg-toast 

Happy holidays :)

jasonfagerberg-toast commented:
@bdemers just realized, this PR was closed and not merged, was that intentional? Mostly just asking when this change will be released
